### PR TITLE
Transfer route ownership

### PR DESCRIPTION
**Pull request by dalvarado**
Co-authored-by: Merric de Launey [mdelauney@pivotal.io](mailto:mdelauney@pivotal.io)
Co-authored-by: Alex Rocha [alexr1@vmware.com](mailto:alexr1@vmware.com)
Co-authored-by: David Alvarado [dalvarado@pivotal.io](mailto:dalvarado@pivotal.io)

Thanks for contributing to cloud_controller_ng. To speed up the process of reviewing your pull request please provide us with:

* A short explanation of the proposed change:

* An explanation of the use cases your change solves

* Links to any other associated PRs

* [ ] I have reviewed the [contributing guide](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/CONTRIBUTING.md)

* [x] I have viewed, signed, and submitted the Contributor License Agreement

* [x] I have made this pull request to the `main` branch

* [x] I have run all the unit tests using `bundle exec rake`

* [ ] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng#cf-acceptance-tests-cats)

### CODE REVIEW

**Comment by Gerg** [comment]
> Note: This is probably note the right endpoint defintion for this action
> 
> current frontrunners are:
> 
> PATCH v3/routes/:guid/relationships/space -d { "data": {"guid": "space-2-guid" }}
> PATCH /v3/routes/:guid/actioins/transfer_owner -d {"guid": "space-2-guid" }
- https://github.com/cloudfoundry/cloud_controller_ng/commit/f3871299c22a8cf0bd17b8a9b86f0b4f495fbf88

What is y'all's current thinking on this?

**Comment by philippthun** [comment]
@dalvarado - Rebasing to the current `main` branch fixes mysql:8.0 tests.

**Comment by MerricdeLauney** [comment]
> > Note: This is probably note the right endpoint defintion for this action
> > current frontrunners are:
> > PATCH v3/routes/:guid/relationships/space -d { "data": {"guid": "space-2-guid" }}
> > PATCH /v3/routes/:guid/actions/transfer_owner -d {"guid": "space-2-guid" }
> 
> Source: [f387129](https://github.com/cloudfoundry/cloud_controller_ng/commit/f3871299c22a8cf0bd17b8a9b86f0b4f495fbf88)
> 
> What is y'all's current thinking on this?

I am still unsure if either of these options (or something else?) are more in align with capi's endpoint design. Would love to hear anyone else's opinion. 

It's worth noting that the current endpoint is  PATCH /v3/routes/:guid/transfer_owner -d {"guid": "space-2-guid" } which is almost certainly not correct.

@Gerg @philippthun

**Comment by Gerg** [comment]
Without thinking about it too deeply, `PATCH v3/routes/:guid/relationships/space` sounds the most intuitive, since you are changing that relationship.

**Comment by philippthun** [comment]
I would also go for the `PATCH ... /relationships/ ...` endpoint.

File: docs/v3/source/includes/resources/routes/_transfer_route_ownership.md.erb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#### Definition
+`PATCH /v3/routes/:route-guid/transfer_owner`
+
+#### Permitted roles
+
+ |
+--- |
+Admin |
+Space Developer |
+Space Supporter |
```
[the remark this answers is not in the document]
Comment (MerricdeLauney) [reply docs/v3/source/includes/resources/routes/_transfer_route_ownership.md.erb:32]: Hmm good question. I think you are right, this falls outside the domain of space supporters

File: app/controllers/v3/routes_controller.rb [hunk 2]
```diff
@@ -167,6 +169,29 @@ def relationships_shared_routes
...
+    unauthorized! unless permission_queryer.can_manage_apps_in_active_space?(route.space.guid)
+
+    target_space = Space.first(guid: message.guid)
+    target_space_error = if target_space.nil? || !can_read_space?(target_space)
+                           'Ensure the space exists and that you have access to it.'
+                         elsif !permission_queryer.can_manage_apps_in_active_space?(target_space.guid)
+                           'Ensure that you have write permission for the target space.'
+                         elsif !permission_queryer.is_space_active?(target_space.guid)
+                           'The target organization is suspended.'
+                         end
```
[the remark this answers is not in the document]
Comment (moleske) [reply app/controllers/v3/routes_controller.rb:188]: started looking at this, still playing around with what making them same looks like.  Such as which parts to change in unshare to match transfer or the opposite way

**Comment by tcdowney** [comment]
Following along with this to see if/when we might want to support this change in Korifi.

> Without thinking about it too deeply, `PATCH v3/routes/:guid/relationships/space` sounds the most intuitive to me, since you are changing that relationship.

Agreed, goes well with the existing route sharing endpoints.
https://v3-apidocs.cloudfoundry.org/version/3.122.0/index.html#share-a-route-with-other-spaces-experimental

```
url "https://api.example.org/v3/routes/[guid]/relationships/shared_spaces" \
  -X POST \
  -H "Authorization: bearer [token]" \
  -H "Content-type: application/json" \
  -d '{
    "data": [
      { "guid": "space-one-guid" },
      { "guid": "space-two-guid" }
    ]
  }'
```

**Pull request merged**